src/implicit.rs

```diff
@@ -27,118 +26,200 @@ pub fn resolve_weak(
+            // (see below for W6 terminator code)
+            //
+            match processing_classes[i] {
+                // <http://www.unicode.org/reports/tr9/#W6>
+                EN => {
                     // W5. If a run of ETs is adjacent to an EN, change the ETs to EN.
                     for j in &et_run_indices {
                         processing_classes[*j] = EN;
                     }
                     et_run_indices.clear();
                 }
-            }
-            // <http://www.unicode.org/reports/tr9/#W3>
-            AL => processing_classes[i] = R,
 
-            // <http://www.unicode.org/reports/tr9/#W4>
-            ES | CS => {
-                // see https://github.com/servo/unicode-bidi/issues/86
-                // We want to make sure we check the correct next character by skipping past the rest
-                // of this one
-                if let Some(ch) = text.get(i..).and_then(|s| s.chars().next()) {
-                    let mut next_class = indices
-                        .clone()
-                        .skip(ch.len_utf8() - 1)
-                        .map(|j| processing_classes[j])
-                        .find(not_removed_by_x9)
-                        .unwrap_or(sequence.eos);
-                    if next_class == EN && last_strong_is_al {
-                        // Apply W2 to next_class. We know that last_strong_is_al
-                        // has no chance of changing on this character so we can still presume its value
-                        next_class = AN;
-                    }
-                    processing_classes[i] = match (prev_class, processing_classes[i], next_class) {
-                        (EN, ES, EN) | (EN, CS, EN) => EN,
-                        (AN, CS, AN) => AN,
-                        (_, _, _) => ON,
+                // <http://www.unicode.org/reports/tr9/#W4>
+                // <http://www.unicode.org/reports/tr9/#W6>
+                ES | CS => {
+                    // See https://github.com/servo/unicode-bidi/issues/86 for improving this.
+                    // We want to make sure we check the correct next character by skipping past the rest
+                    // of this one.
+                    if let Some(ch) = text.get(i..).and_then(|s| s.chars().next()) {
+                        let mut next_class = sequence
+                            .iter_forwards_from(i + ch.len_utf8(), run_index)
+                            .map(|j| processing_classes[j])
+                            // <https://www.unicode.org/reports/tr9/#Retaining_Explicit_Formatting_Characters>
+                            .find(not_removed_by_x9)
+                            .unwrap_or(sequence.eos);
+                        if next_class == EN && last_strong_is_al {
+                            // Apply W2 to next_class. We know that last_strong_is_al.
+                            // Has no chance of changing on this character so we can still presume its value.
```

Comment:
_presume_ feels like the wrong verb here.
We can preserve/reuse/retain its value? Or just we do not need to update it?